Mathlib/Tactic/FunProp/Types.lean

```diff
@@ -66,49 +63,71 @@ def FunctionData.getFnOrigin (fData : FunctionData) : Origin :=
   | .const name _ => .decl name
   | _ => .decl Name.anonymous
 
+/-- Error logging mode for `fun_prop`. Main mode is used for messages that should be directly
+displayed to the user and secondary mode is for messages that are displayed by setting option
+`Meta.Tactic.fun_prop.errors` to true. -/
+inductive LoggingMode where
+  /-- Main error logging mode, these messages will be diplayed to the user be default. -/
+  | main
+  /-- Secondary error logging mode, these messages will be diplayed to the user by setting option
+  `Meta.Tactic.fun_prop.errors` to true . -/
+  | secondary
+
 /-- Default names to be considered reducible by `fun_prop` -/
 def defaultNamesToUnfold : Array Name :=
   #[`id, `Function.comp, `Function.HasUncurry.uncurry, `Function.uncurry]
 
 /-- `fun_prop` configuration -/
 structure Config where
+  /-- Maximal number of transitions between function properties
+  e.g. inferring differentiability from linearity -/
+  maxDepth := 200
+  /-- Maximum number of steps `fun_prop` can take. -/
+  maxSteps := 100000
+  /-- Use transition theorem. -/
+  useTransThms := true
+deriving Inhabited
+
+/-- `fun_prop` context -/
+structure Context where
+  /-- fun_prop config -/
+  config : Config := {}
   /-- Name to unfold -/
   constToUnfold : Batteries.RBSet Name Name.quickCmp :=
     .ofArray defaultNamesToUnfold _
   /-- Custom discharger to satisfy theorem hypotheses. -/
   disch : Expr → MetaM (Option Expr) := fun _ => pure .none
-  /-- Maximal number of transitions between function properties
-  e.g. inferring differentiability from linearity -/
-  maxDepth := 200
   /-- current depth -/
   depth := 0
   /-- Stack of used theorem, used to prevent trivial loops. -/
   thmStack : List Origin := []
-  /-- Maximum number of steps `fun_prop` can take. -/
-  maxSteps := 100000
-deriving Inhabited
+  /-- Logging mode, determines whether logged messages will be diplayes to the user immediately or
+  on request by setting `Meta.Tactic.fun_prop.errors` to true. -/
+  loggingMode : LoggingMode := .main
 
 /-- `fun_prop` state -/
 structure State where
-  /-- Simp's cache is used as the `funProp` tactic is designed to be used inside of simp and utilize
-  its cache -/
-  cache : Simp.Cache := {}
+  /-- Simp's cache is used as the `fun_prop` tactic is designed to be used inside of simp and
+  utilize its cache -/
+  cache        : Simp.Cache := {}
   /-- Count the number of steps and stop when maxSteps is reached. -/
   numSteps := 0
   /-- Log progress and failures messages that should be displayed to the user at the end. -/
-  msgLog : List String := []
+  mainMsgLog : List String := []
+  /-- Log progress and failures messages that should be displayed to the user at the end when
+  the option `Meta.Tactic.fun_prop.errors` is set to true. -/
+  secondaryMsgLog : List String := []
```

Comment:
This entire approach to logging seems a bit peculiar to me. I feel like the more common ways to do this would be either:

### Option A

Use just one `msgLog` as before and no "LoggingMode". Create an option

```
register_option Meta.Tactic.fun_prop.errors : Bool := {
  defValue := false
  descr := "yada yada" }
```

and then use `if Meta.Tactic.fun_prop.errors.get (← getOptions) then` whenever you want to add a "secondary" log message. (this approach does not sort them as main/secondary, but seems much simpler)

### Option B

Just use `trace[Meta.Tactic.fun_prop.errors]` wherever you have these secondary error messages, don't bother about the sorting and have the error say something like "enable `set_option trace.Meta.Tactic.fun_prop.errors true` to see more details!" No need for the "LoggingMode" either.

### Option C
If you really think adding `LoggingMode` adds value, I'd still suggest you use `register_option` instead of `registerTraceClass`.

In any case, it would be nice to have a test in the dev-test file using `#guard_msgs` which demonstrates the expected error message, would help understanding and reviewing..

Reply:
The secondary errors are already reported with `set_option trace.Meta.Tactic.fun_prop` so I'm going to completely ditch the main/secondary dichotomy. 

I'm also going to include few tests with `#guard_msgs` to the dev file as you suggest.

Reply:
Thanks!